### PR TITLE
hyperkit: init at 0.20190201

### PR DESCRIPTION
**Pull request by nicknovitski**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- ~[ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))~
- ~[ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`~
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- ~[ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)~
- ~[ ] Ensured that relevant documentation is up to date~
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/virtualization/hyperkit/default.nix [hunk 1]
```diff
@@ -0,0 +1,51 @@
+{ stdenv, lib, fetchFromGitHub
+, Hypervisor, vmnet, SystemConfiguration
+, xpc, libobjc, dtrace
+}:
+
+let
+  # Make sure to keep those in sync
```
Comment (mmahut) [inline pkgs/applications/virtualization/hyperkit/default.nix]: These are not in sync. Tag ```v0.20190201``` referes to ```64bbfbec66900334d91bb2ac32de60b5187cf6c9```.

Question, why not only using the tag itself?

Comment (nicknovitski) [reply pkgs/applications/virtualization/hyperkit/default.nix]: The tool's `--version` option includes both in its response.  [The Makefile uses git to determine them dynamically](https://github.com/moby/hyperkit/blob/master/Makefile#L1-L2), but you can see the prePatch phase below which removes that usage.

But we could just use `v${version}` for both and make it impossible for the version and rev to fall out of sync.

Comment (nicknovitski) [reply pkgs/applications/virtualization/hyperkit/default.nix]: Actually, I'm wrong, compilation does require a CFLAG `-DVERSION_SHA1`, but I can't tell what it's used for.
```
% result/bin/hyperkit -v
hyperkit: v0.20190201

Homepage: https://github.com/docker/hyperkit
License: BSD
```

Comment (nicknovitski) [reply pkgs/applications/virtualization/hyperkit/default.nix]: I've updated the PR to use the tag for every version and "sha" string.

Comment (nicknovitski) [reply pkgs/applications/virtualization/hyperkit/default.nix]: I was wrong, that won't work, [there's a runtime error if the VERSION_SHA1 is not the right length](https://github.com/moby/hyperkit/blob/3e954c1068f2ebea39cbc0119fe4d0f31f015f3a/src/lib/pci_virtio_net_vpnkit.c#L324)

**Comment by nicknovitski** [comment]
Are there any other changes I should make?

File: pkgs/applications/virtualization/hyperkit/default.nix [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  buildInputs = [ Hypervisor vmnet SystemConfiguration xpc libobjc dtrace ];
+
+  # 1. Don't use git to determine version
+  # 2. Include dtrace probes
+  prePatch = ''
+    substituteInPlace Makefile \
+      --replace 'shell git describe --abbrev=6 --dirty --always --tags' "v${version}" \
+      --replace 'shell git rev-parse HEAD' "${src.rev}" \
+      --replace 'PHONY: clean' 'PHONY:'
```
Comment (Enteee) [inline pkgs/applications/virtualization/hyperkit/default.nix]: Why are you are making the `clean` target non-phony here?

File: pkgs/applications/virtualization/hyperkit/default.nix [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    make src/include/xhyve/dtrace.h
```
Comment (Enteee) [inline pkgs/applications/virtualization/hyperkit/default.nix]: Won't this automatically happen during the build anyways (https://github.com/moby/hyperkit/blob/97f091f9a65390123d96c6794a4b29190e04ce3d/Makefile#L158), so is this needed here?

File: pkgs/applications/virtualization/hyperkit/default.nix [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    mkdir -p $out/bin
+    cp build/hyperkit $out/bin
+  '';
+
+  meta = with lib; {
+    description = "A toolkit for embedding hypervisor capabilities in your application";
+    homepage = https://github.com/moby/hyperkit;
+    maintainers = with maintainers; [ nicknovitski ];
+    platforms = platforms.darwin;
+    license = licenses.bsd3;
```
Comment (Enteee) [inline pkgs/applications/virtualization/hyperkit/default.nix]: as far as I can see, their current license (from the README.md) looks more like bsd2.

**Comment by abathur** [comment]
I'm not sure how big the gap is here now, especially over time (nor do I know anything about the software--I just found this PR when I had to go install hyperkit as a dependency to try something else out), but it seems like a shame that this got bogged down.

**Review by SuperSandro2000 — COMMENTED** [review]
Is this identical to pkgs/applications/networking/cluster/docker-machine/hyperkit.nix?

**Comment by abathur** [comment]
> Is this identical to pkgs/applications/networking/cluster/docker-machine/hyperkit.nix?

It looks like that (docker-machine/hyperkit.nix) is a docker-machine driver for this. I'll rephrase in relation to homebrew's formula, since they appear to have both:
- docker-machine/hyperkit.nix looks like an analogue of https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/docker-machine-driver-hyperkit.rb 
- pkgs/applications/virtualization/hyperkit/default.nix introduced here looks like the analogue of https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/hyperkit.rb

**Comment by SuperSandro2000** [comment]
> I'm not sure how big the gap is here now, especially over time (nor do I know anything about the software--I just found this PR when I had to go install hyperkit as a dependency to try something else out), but it seems like a shame that this got bogged down.

You can pick it up if you want.

**Comment by SuperSandro2000** [comment]
Closing due to inactivity from author.

**Pull request closed without merging**